### PR TITLE
Add Scala Interface

### PR DESCRIPTION
**Pull request by abhinavrai44**
Issue #3485

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1104,6 +1105,18 @@ IF (PerlModular)
 	ENDIF()
 ENDIF()
 
+# Scala modular
+IF (ScalaModular)
+	FIND_PACKAGE(Scala REQUIRED)
+	UNSET(TARGET_SWIGFLAGS)
+	IF(EXISTS ${CMAKE_SOURCE_DIR}/src/interfaces/java_modular)
+		IF(!JavaModular)
```
Comment (karlnapf) [inline CMakeLists.txt]: This is so that you can use the java files I guess. Good

File: cmake/ShogunPackaging.cmake [hunk 2]
```diff
@@ -88,6 +88,7 @@ else()
 	set(CPACK_COMPONENT_java_DISPLAY_NAME "Java interface")
 	set(CPACK_COMPONENT_octave_DISPLAY_NAME "OCTAVE interface")
 	set(CPACK_COMPONENT_lua_DISPLAY_NAME "Lua interface")
+	set(CPACK_COMPONENT_lua_DISPLAY_NAME "Scala interface")
```
Comment (karlnapf) [inline cmake/ShogunPackaging.cmake]: copy paste typo

File: CMakeLists.txt (hunk 1 above)
Comment (abhinavrai44) [inline CMakeLists.txt]: Yeah, cause scala can directly use Java files. So I think we don't need another modular for Scala

File: examples/meta/generator/targets/scala.json [hunk 3]
```diff
@@ -0,0 +1,99 @@
+{
```
Comment (karlnapf) [inline examples/meta/generator/targets/scala.json]: I think we don't need another json here, just use the existing java json, the syntax is example the same

File: examples/meta/scala/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+SET(SCALAJAR "/usr/share/scala-2.11/lib/scala-library.jar")
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: this has to be detected by cmake
(send another PR for detecting scala)

File: examples/meta/scala/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+SET(CLASSPATH "${JBLAS}:${SHOGUN_JAR}:${SCALAJAR}:${CMAKE_CURRENT_BINARY_DIR}")
+SET(JAVAOPTS "-Xmx1024m")
+SET(JAVA_LIB_PATH "${JAVA_MODULAR_BUILD_DIR}")
+
+SET(SCALA_SCALAC_EXECUTABLE "/usr/bin/scalac")
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: same here, do that in a sep patch

File: examples/meta/scala/CMakeLists.txt (hunk 4 above)
Comment (abhinavrai44) [inline examples/meta/scala/CMakeLists.txt]: Ok, right I was hard coding it for testing it on my system

File: examples/meta/scala/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	    get_filename_component(FULL_DIR ${META_EXAMPLE} PATH)
+	ELSE()
+	    # only supported from >2.8.11
+	    get_filename_component(FULL_DIR ${META_EXAMPLE} DIRECTORY)
+	ENDIF()
+	get_filename_component(EXAMPLE_REL_DIR ${FULL_DIR} NAME)
+	set(EXAMPLE_NAME_WITH_DIR "${EXAMPLE_REL_DIR}-${EXAMPLE_NAME}")
+
+	ADD_CUSTOM_COMMAND(OUTPUT compiled_scala_${EXAMPLE_NAME_WITH_DIR}
+				  COMMAND ${SCALA_SCALAC_EXECUTABLE} -cp ${CLASSPATH}
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: I think the environmental variables here are mixed up with java a bit?

File: cmake/ShogunPackaging.cmake (hunk 2 above)
Comment (abhinavrai44) [inline cmake/ShogunPackaging.cmake]: Sorry

**Comment by abhinavrai44** [comment]
@karlnapf Scala can't compile java files but it can run the class files generated.
https://www.scala-lang.org/old/node/5248.html

So what do you think, how should I proceed ?

**Comment by karlnapf** [comment]
Yes, so let's just execute the compiled java examples using scala then.

File: .travis.yml [hunk 7]
```diff
@@ -60,6 +60,12 @@ matrix:
         - CMAKE_OPTIONS="-DRModular=ON -DTRAVIS_DISABLE_UNIT_TESTS=ON -DTRAVIS_DISABLE_LIBSHOGUN_TESTS=ON"
         - CC=clang
         - CXX=clang++
+    - compiler: clang
+      services: docker
+      env:
+        - CMAKE_OPTIONS="-DScalaModular=ON -DTRAVIS_DISABLE_UNIT_TESTS=ON -DTRAVIS_DISABLE_LIBSHOGUN_TESTS=ON"
+        - CC=clang
+        - CXX=clang++
```
Comment (vigsterkr) [inline .travis.yml]: this aint' going to work until you dont add scala runtime to the shogun docker sdk

File: configs/shogun-sdk/Dockerfile [hunk 8]
```diff
@@ -33,3 +33,8 @@ RUN cd /tmp;\
     cd arprec-2.2.18;\
     ./configure --enable-shared;\
     make install; ldconfig
+
+RUN wget www.scala-lang.org/files/archive/scala-2.11.7.deb
+RUN sudo dpkg -i scala-2.11.7.deb
+RUN apt-get update -qq
```
Comment (vigsterkr) [inline configs/shogun-sdk/Dockerfile]: there's absolutely no need to do update and apt-get install scala; the latter would actually would install trusty's scala 2.9...

Comment (abhinavrai44) [inline configs/shogun-sdk/Dockerfile]: So, I only need to keep the first 2 lines and it will be fine ?

File: .travis.yml (hunk 7 above)
Comment (karlnapf) [inline .travis.yml]: which needs a sep PR, I suggest do modify docker config first (it auto-updates the image when we merge a PR to the config file), and then this travis change in a single PR.
Then the scala stuff you made in a next PR

File: examples/meta/generator/targets/scala.json (hunk 3 above)
Comment (karlnapf) [inline examples/meta/generator/targets/scala.json]: As said before, this file can go. We just want to execute the already present java files

**Comment by karlnapf** [comment]
Lets get this moving! :)

**Comment by abhinavrai44** [comment]
Yeah, right on it.

File: examples/meta/generator/targets/scala.json (hunk 3 above)
Comment (abhinavrai44) [inline examples/meta/generator/targets/scala.json]: If the json file is removed the corresponding folder binary_classifier, distance, etc in build/tests/meta/generated_results/scala are not created. And therefore the .dat files for scala for each category cannot be saved.

**Comment by vigsterkr** [comment]
ok this needs rebase, because of my refactor of cmake but as well to see how it actually works now that our shogun-dev has scala in it and the scala detector is already merged...

File: CMakeLists.txt [hunk 9]
```diff
@@ -528,6 +529,16 @@ IF (PerlModular)
 	ENDIF()
 ENDIF()
 
+# scala modular
+IF (ScalaModular)
+	FIND_PACKAGE(Scala REQUIRED)
+	IF(NOT JavaModular)
+		IF(EXISTS ${CMAKE_SOURCE_DIR}/src/interfaces/java_modular)
+			add_subdirectory(${CMAKE_SOURCE_DIR}/src/interfaces/java_modular)
```
Comment (karlnapf) [inline CMakeLists.txt]: please put an info message telling the user that java is enabled because scala needs that

File: examples/meta/generator/generate.py [hunk 10]
```diff
@@ -123,6 +123,7 @@ def translateExamples(inputDir, outputDir, targetsDir, ctagsFile,
         'csharp',
         'ruby',
         'lua',
+        'scala',
```
Comment (karlnapf) [inline examples/meta/generator/generate.py]: I thought we don't generate scala code? But rather run compiled java examples using scala ?

File: examples/meta/generator/targets/scala.json (hunk 3 above)
Comment (karlnapf) [inline examples/meta/generator/targets/scala.json]: ẁe should maybe then create these folders by hand

Comment (karlnapf) [inline examples/meta/generator/targets/scala.json]: Definitely no json file please

File: examples/meta/generator/translate.py [hunk 11]
```diff
@@ -642,7 +642,7 @@ def loadTargetDict(targetJsonPath):
     parser.add_argument("-t",
                         "--target",
                         nargs='?',
-                        help="Translation target. Possible values: cpp, python, java, r, octave, csharp, ruby, lua. (default: python)")
+                        help="Translation target. Possible values: cpp, python, java, r, octave, csharp, ruby, lua, scala. (default: python)")
```
Comment (karlnapf) [inline examples/meta/generator/translate.py]: no scala pls

File: examples/meta/scala/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+	get_filename_component(EXAMPLE_REL_DIR ${FULL_DIR} NAME)
+	set(EXAMPLE_NAME_WITH_DIR "${EXAMPLE_REL_DIR}-${EXAMPLE_NAME}")
+
+	ADD_CUSTOM_COMMAND(OUTPUT compiled_scala_${EXAMPLE_NAME_WITH_DIR}
+				  COMMAND ${Java_JAVAC_EXECUTABLE} -cp ${CLASSPATH}
+					-d ${CMAKE_CURRENT_BINARY_DIR}/${EXAMPLE_REL_DIR} ${CMAKE_CURRENT_BINARY_DIR}/${EXAMPLE_REL_DIR}/${EXAMPLE_NAME}.java
+				  DEPENDS java_modular)
+	LIST(APPEND GENERATED_SCALA_EXAMPLES compiled_scala_${EXAMPLE_NAME_WITH_DIR})
+
+	add_test(NAME generated_scala-${EXAMPLE_NAME_WITH_DIR}
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: please dont compile scala examples, you just need to *run* the compiled java examples

File: tests/meta/CMakeLists.txt [hunk 13]
```diff
@@ -64,6 +64,7 @@ FOREACH(REFERENCE_FILE ${META_INTEGRATION_REFERENCES})
 	            AddMetaIntegrationTest(octave OctaveModular)
 	            AddMetaIntegrationTest(csharp CSharpModular)
 	            AddMetaIntegrationTest(ruby RubyModular)
+	            AddMetaIntegrationTest(scala ScalaModular)
```
Comment (karlnapf) [inline tests/meta/CMakeLists.txt]: this should stay in as we definitely want to integration test the scala output

File: CMakeLists.txt (hunk 9 above)
Comment (abhinavrai44) [inline CMakeLists.txt]: ok.

File: examples/meta/generator/targets/scala.json (hunk 3 above)
Comment (abhinavrai44) [inline examples/meta/generator/targets/scala.json]: Ok. I'll work on some workaround

File: examples/meta/scala/CMakeLists.txt (hunk 12 above)
Comment (abhinavrai44) [inline examples/meta/scala/CMakeLists.txt]: Got it. :-)

File: examples/meta/generator/targets/scala.json (hunk 3 above)
Comment (karlnapf) [inline examples/meta/generator/targets/scala.json]: you can do that with cmake btw

**Comment by karlnapf** [comment]
How are things going?

**Comment by abhinavrai44** [comment]
Done. Just one small change remaining. Will commit in a short while :-)

**Comment by abhinavrai44** [comment]
@karlnapf I have made the changes. Plz check ..

File: doc/cookbook/source/conf.py [hunk 14]
```diff
@@ -208,6 +208,7 @@
         ('lua', 'lua'),
         ('csharp', 'cs'),
         ('cpp', 'cpp'),
+        ('scala','scala'),
```
Comment (karlnapf) [inline doc/cookbook/source/conf.py]: Maybe we can leave the cookbook out for now, and add scala there in a second attempt (we can just rename the tab to "java/scala" in fact)

File: doc/cookbook/source/conf.py [hunk 15]
```diff
@@ -219,6 +220,7 @@
                 'ruby': 'Ruby',
                 'cpp': 'Native C++',
                 'lua': 'Lua',
+                'scala': 'Scala',
```
Comment (karlnapf) [inline doc/cookbook/source/conf.py]: Rather change java to "java/scala"

File: examples/meta/generator/targets/java.json [hunk 16]
```diff
@@ -1,5 +1,5 @@
 {
-    "Program": "import org.jblas.DoubleMatrix;\nimport org.jblas.FloatMatrix;\n\nimport org.shogun.modshogun;\n$dependencies\n\nclass $programName {\nstatic {\nSystem.loadLibrary(\"modshogun\");\n}\n\npublic static void main(String argv[]) {\nmodshogun.init_shogun_with_defaults();\n\n$program\n}\n}\n",
+    "Program": "import org.jblas.DoubleMatrix;\nimport org.jblas.FloatMatrix;\n\nimport org.shogun.modshogun;\n$dependencies\n\npublic class $programName {\nstatic {\nSystem.loadLibrary(\"modshogun\");\n}\n\npublic static void main(String argv[]) {\nmodshogun.init_shogun_with_defaults();\n\n$program\n}\n}\n",
```
Comment (karlnapf) [inline examples/meta/generator/targets/java.json]: why this?

File: examples/meta/scala/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+			WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${EXAMPLE_REL_DIR}
+			COMMAND ${Scala_SCALA_EXECUTABLE}
+				-cp ${CLASSPATH}:${CMAKE_JAVA_DIR}/${EXAMPLE_REL_DIR}
+				-Djava.library.path=${JAVA_LIB_PATH} ${EXAMPLE_NAME})
+
+ENDFOREACH()
+
+add_custom_target(build_scala_meta_examples ALL
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: this target doesnt make sense if there is not scala compiled code right?

File: examples/meta/scala/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+	ENDIF()
+	get_filename_component(EXAMPLE_REL_DIR ${FULL_DIR} NAME)
+	set(EXAMPLE_NAME_WITH_DIR "${EXAMPLE_REL_DIR}-${EXAMPLE_NAME}")
+	set(SCALA_FOLDER "${CMAKE_CURRENT_BINARY_DIR}/${EXAMPLE_REL_DIR}")
+
+	file(MAKE_DIRECTORY ${SCALA_FOLDER})
+
+	add_test(NAME generated_scala-${EXAMPLE_NAME_WITH_DIR}
+			WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${EXAMPLE_REL_DIR}
+			COMMAND ${Scala_SCALA_EXECUTABLE}
```
Comment (karlnapf) [inline examples/meta/scala/CMakeLists.txt]: cool! this works on your machine?

File: tests/meta/CMakeLists.txt (hunk 13 above)
Comment (karlnapf) [inline tests/meta/CMakeLists.txt]: Actually, I think we don't need integration tests for scala, see other comment

**Comment by karlnapf** [comment]
Cool this looks good apart from some minor issues....

@vigsterkr I guess we don't need integration testing of the meta examples for scala (since that is already done for java, and I trust them that the code does the same)

File: examples/meta/generator/targets/java.json (hunk 16 above)
Comment (abhinavrai44) [inline examples/meta/generator/targets/java.json]: Scala gives the error -  cannot access a member of class with modifiers public static. To solve this the class had to be made public

File: examples/meta/scala/CMakeLists.txt (hunk 18 above)
Comment (abhinavrai44) [inline examples/meta/scala/CMakeLists.txt]: Yes it works. I'll just make the suggested changes and resubmit.

**Comment by karlnapf** [comment]
Actually there is one more thing: the integration tests for java need to be executed using the `*.dat` files that java generated. As you are executing the java compiled examples in their directory, these are overwritten by the scala ones....

Solution 1: make the java examples depend on the scala ones so that scala is executed before java.
Solution 2: Replicate the example folder structure for scala and symlink the example binaries in there, then execute in the new folder. This also gives the option to do seperate integration testing, although I think we dont need that

@vigsterkr thoughts?

File: examples/meta/generator/targets/java.json (hunk 16 above)
Comment (karlnapf) [inline examples/meta/generator/targets/java.json]: OK! thanks for the explain

**Comment by karlnapf** [comment]
BTW please also add a NEWS entry that says that we now have scala added to our build

File: tests/meta/CMakeLists.txt (hunk 13 above)
Comment (karlnapf) [inline tests/meta/CMakeLists.txt]: Finally, I changed my mind again, seeing how nicely travis executes the integration tests. So leave it in, it is redundant but I think its ok to have it

**Comment by karlnapf** [comment]
The summary of my messy review: This is good to be merged once you removed the cookbook extension and changed the tab title

File: doc/cookbook/source/conf.py (hunk 15 above)
Comment (abhinavrai44) [inline doc/cookbook/source/conf.py]: Ok.

File: examples/meta/scala/CMakeLists.txt (hunk 17 above)
Comment (abhinavrai44) [inline examples/meta/scala/CMakeLists.txt]: Yeah. I'll remove this

File: doc/cookbook/source/conf.py (hunk 14 above)
Comment (karlnapf) [inline doc/cookbook/source/conf.py]: This one can be removed

File: doc/cookbook/source/conf.py (hunk 15 above)
Comment (karlnapf) [inline doc/cookbook/source/conf.py]: This is the tab title thing

**Comment by karlnapf** [comment]
Check for example http://shogun.ml/examples/latest/examples/binary_classifier/kernel_svm.html
It has these language tabs at the top

**Comment by abhinavrai44** [comment]
Yeah ok..

**Comment by abhinavrai44** [comment]
@karlnapf Done....

File: NEWS [hunk 19]
```diff
@@ -1,3 +1,8 @@
+2017-02-17 Abhinav Rai <abhinavrai44@gmail.com>
+	* SHOGUN Release version 5.0.0 (libshogun 17.2, data 0.11, parameter 1)
+	* Features:
```
Comment (karlnapf) [inline NEWS]: You need to put just one line in the 5.1 here...no need to create a new release ;)
Also just say that we added Scala interface to the build, it's not like we added a new language ...

File: doc/cookbook/source/conf.py [hunk 20]
```diff
@@ -214,7 +214,7 @@
                 'python': 'Python',
                 'octave': 'Octave',
                 'r': 'R',
-                'java': 'Java',
+                'java': 'Java/Scala',
```
Comment (karlnapf) [inline doc/cookbook/source/conf.py]: Yep!!

**Comment by karlnapf** [comment]
Cool then :) Nice work!
